lib/cyrusdb_skiplist.c

```diff
@@ -250,7 +250,7 @@ static int myinit(const char *dbdir, int myflags)
 
         if (r != -1) r = ftruncate(fd, 0);
         net32_time = htonl(global_recovery);
-        if (r != -1) r = write(fd, &net32_time, 4);
+        if (r != -1) r = retry_write(fd, &net32_time, 4);
```

Comment:
I am going to approve this PR but I find use of the `r` variable in this function quite confusing.  I know you just built on what's already there (except for storing a file descriptor in `r`), but maybe you are up to clean that variable up, too.

Some points I find confusing:

- typically, we use `r` as the return value for a function. But this function hardcodes its return values to either be 0 or CYRUSDB_IOERROR, so `r` might already be a misnomer.
- for the read() and retry_write() calls, `r` is meant to store a `ssize_t`, whereas we know that exactly 4 bytes must be read or written or otherwise it's an error
- `r` now also stores the file descriptor `fd` value, whereas the code could just as well check `fd !- -1` which might make it even more clear on what's the error condition.
- in case of ftruncate() `r` stores an (inversed) boolean, but at least we know that false is -1 here.

Reply:
I did think about and tinker with a lot of the things you raise, but they mostly wound up cluttering things up and making it harder, not easier, to read.

We do often use `r` like this, to keep track of an internal success or failure state but eventually returning some other success or failure value (or just void), so I don't think this use of the name `r` is unusual.  We do also use it the way you describe, and probably some other ways too.  I think of `r` as a place for storing the result of some function call (but not necessarily the eventual result of _this_ function call).  Sometimes we also return it up to our caller, but not always.

For what it's worth, `retry_read(..., n)` and `retry_write(..., n)` only ever return `-1` or `n`.  Internally, they repeat the syscall until they've successfully read or written the expected number of bytes.  If anything goes wrong, they return `-1`, _not_ the count-so-far.  When dealing with nontrivial files, that can make it tricky to reason about the current file offset.  But for the kind of thing we're doing here, the binary it-either-completely-succeeded-or-it-failed behaviour is the point, and is why I don't need to also check for `4`.